### ISSUE TITLE
VLAN support?

### ISSUE THREAD

**Issue by Nowatel** [xdp-project/xdp-cpumap-tc#7]
Hi,

Is there a way to support vlan's via xdp-cpumap-tc? As I understand we can only redirect the traffic to the physical nic like eth0 as they have xps_cpus.  Virtual interfaces like eth0.123 don't have multiple xps_cpus that we can use. I didn't see any example involving shaping on the vlan via this method.

**Comment by interduo** [comment]
Hi @Nowatel did You find the answer for that?

**Comment by Nowatel** [comment]
> Hi @Nowatel did You find the answer for that?

Unfortunately it's not possible. As referenced in mentioned topic it can only be done via additional libreqos box.

**Comment by tohojo** [comment]
There is already VLAN support. The packet parsing code already handles (i.e., skips) VLAN tags when it encounters them: https://github.com/xdp-project/xdp-cpumap-tc/blob/master/src/xdp_iphash_to_cpu_kern.c#L99-L120

If you just run the XDP program on the physical NIC (i.e., **not** the virtual VLAN interface), you'll get the packets redirected to the right CPU based on the inner IP header, and the network stack will take over on that CPU afterwards, handling the VLAN tagging and routing the packets to the stacked VLAN interface.

So it should be quite possible to use this as-is even in a VLAN setup...

**Comment by netoptimizer** [comment]
> There is already VLAN support. 

Yes, even double-tagged VLAN support (which is run in production at @danielnielsennumber1 ).

As @thebracket noticed in rchac/LibreQoS#26 the catch is that you need to disable hardware VLAN offloading on the NIC.
This is needed for XDP, as XDP need to vlan headers inline in the packet (in the future with XDP-hints this will be resolved).

Use commands:
```
 ethtool -K (interface-out) rxvlan off
 ethtool -K (interface-in) rxvlan off
```

The setup script [bin/set_irq_affinity_with_rss_conf.sh](https://github.com/xdp-project/xdp-cpumap-tc/blob/master/bin/set_irq_affinity_with_rss_conf.sh) also [disable](https://github.com/xdp-project/xdp-cpumap-tc/blob/12ea5973bac5/bin/set_irq_affinity_with_rss_conf.sh#L129) hardware offloaded VLAN.

**Comment by interduo** [comment]
@netoptimizer So using the same card with two VLANs as inbound and outbound interface is possible? (case described here: https://github.com/rchac/LibreQoS/issues/48)

**Comment by thebracket** [comment]
I haven't tested it, but it really should work. Make your parent bridge
contain the two VLANs (and not the parent interface), and use those as the
two interfaces. The packets coming in will still be seen by the right
xdp/tc programs, which will figure out where the payload is relative to the
tags - so I can't think of a good reason why it wouldn't work. NO idea how
well that would perform.

**Comment by Strykar** [comment]
Quick note for anyone watching this thread: the `parse_eth` block linked above (`xdp_iphash_to_cpu_kern.c`#L99-L120) hand-unrolls exactly two VLAN unwraps, so frames with 3 or more tags fall through to `XDP_PASS` without ever hitting the per-IP CPU classification path. The bypass doesn't show up at the TC layer because of kernel VLAN normalization, details are in the PR verification section.

PR #15 raises the limit to `VLAN_MAX_DEPTH` (default 4), matching the xdp-tools `parse_ethhdr()` convention. The double-tagged production setup @netoptimizer mentioned upthread is unaffected, that's 2 tags, which the existing code handles correctly. The bypass only kicks in at 3+ tags. cc @danielnielsennumber1 since this likely affects your deployment.
